src/NHibernate/Impl/AbstractSessionImpl.cs

```diff
@@ -293,6 +292,50 @@ protected bool IsAlreadyDisposed
 
 		public abstract bool TransactionInProgress { get; }
 
+		/// <inheritdoc/>
+		public bool AutoFlushSuspended => _suspendAutoFlushCount != 0;
+
+		[NonSerialized]
+		private int _suspendAutoFlushCount;
+
+		/// <inheritdoc/>
+		public IDisposable SuspendAutoFlush()
+		{
+			return new SuspendAutoFlushHelper(this);
+		}
+
+		private class SuspendAutoFlushHelper : IDisposable
+		{
+			private AbstractSessionImpl _session;
+
+			public SuspendAutoFlushHelper(AbstractSessionImpl session)
+			{
+				_session = session;
+				_session._suspendAutoFlushCount++;
+			}
+
+			#region IDisposable Support
+			private bool _disposedValue; // To detect redundant calls
+
+			protected virtual void Dispose(bool disposing)
```

Comment:
This class is private, so I dont see the reasons to have virtual members and/or implement Disposable "correctly"

Reply:
Default standard snippet from VS, including the anti-redundant call (which is a must for this case). 

And well, we could consider than when `disposing` is false (meaning being called by GC, explicit dispose lacking) we should left the counter incremented for avoiding "fixing" the lack of dispose call through GC and hiding somewhat it.

Reply:
Now moved to `SessionImpl`, pointless `virtual` removed.

Reply:
But, then we miss finilizer, which should call `Dispose(false)` then

Reply:
No more needed since Fx 2 I believe, handled by GC without any finalizer. The fact that `SessionImpl` still do that is an old remnant of Fx 1.

Reply:
But anyway no one is going to call `Dispose(false)`, so I don't think we need to follow this obsolete pattern. Also, I don't think that framework will call `Dispose(false)` or `Dispose()` on the finalization stage. We need to prevent calling `Dispose()` twice, but anyway, without this check we'll get `NullReferenceException` as session would be unset by the time of second time `Dispose()` is called.

Reply:
From where is the `protected virtual Dispose(boolean disposing)` pattern said to be obsolete? Only implementing explicitly a finalizer is obsolete for most cases. This [doc](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose) seems quite current.

Now for such a little class it should not ever have consequences, but I just rather stick to recommended patterns when they are strongly recommended as in previously linked documentation.

Reply:
I don't see where it says that finalize is optional:

What I see is:
> You **must** override Finalize to ensure that unmanaged resources are disposed of if your System.IDisposable.Dispose implementation is not called by a consumer of your type.

But again, we implement the `IDisposable` interface not for it's purpose, but because of convenience of `using`.

Reply:
The fact is my implementation is invalid for the disposing pattern because I should not touch the managed session when disposing is false: if both were dereferenced, the session could have been GC already and should no more be used in any way.

I am probably wrong by telling the GC call disposing(false) itself, it is likely only useful if some later inheritor decide to implement its own finalizer though this should no more be the way to go (use SafeHandle instead for not having to handle directly the releasing of unmanaged resources).

But well, following the recommended practice just avoids having to think about such details while not being a significant overhead.

Reply:
> This is the recommended variation and doesn't require overriding the System.Object.Finalize method.

This is in the first point above. The "must" is in the second way of implementing dispose, while the recommended way is the first.